hamilton/plugins/jupyter_magic.py

```diff
@@ -77,159 +131,272 @@ def insert_cell_with_content():
-        module_name = self.get_module_name(args)
+        args, unknown_args = parse_known_argstring(
+            self.cell_to_module, line
+        )  # specify how to parse by passing method
+        self.resolve_unknown_args_cell_to_module(unknown_args)
+
+        # validate variables exist in the user namespace expect `config` because it's a special case
+        # will exit using `return` in case of error
+        args_that_read_user_namespace = ["display", "builder", "inputs", "overrides"]
+        for name, value in vars(args).items():
+            if name not in args_that_read_user_namespace:
+                continue
+
+            # special case: `display` can be passed as a flag (=True), without a config
+            if name == "display" and value is True:
+                continue
+
+            # main case: exit if variable is not in user namespace
+            if value and self.shell.user_ns.get(value) is None:
+                return f"KeyError: Received `--{name} {value}` but variable not found."
+
+        # special case: `config` expects potentially a JSON string
+        # there for backwards compatibility
+
+        # default case: didn't receive `-c/--config`. Set an empty dict
+        if args.config is None:
+            config = {}
+        # case 1, 2, 3: `-c/--config` is specified
+        # case 1: -c/--config refers to variable in the user namespace
+        elif self.shell.user_ns.get(args.config):
+            config = self.shell.user_ns.get(args.config)
+        # case 2: parse using key=value
+        elif "=" in args.config:
+            config = parse_config(args.config)
+        # case 3: parse as JSON
+        elif ":" in args.config:
+            try:
+                # strip quotation marks added by IPython and avoid mutating `args`
+                config_str = args.config.strip("'\"")
+                config = json.loads(config_str)
+            except json.JSONDecodeError:
+                print(f"JSONDecodeError: Failed to parse `config` as JSON. Received {value}")
+                return
 
-        display_config = self.get_display_config(args)
+        # get the default values of args
+        module_name = args.module_name
+        base_builder = self.shell.user_ns[args.builder] if args.builder else driver.Builder()
+        inputs = self.shell.user_ns[args.inputs] if args.inputs else {}
+        overrides = self.shell.user_ns[args.overrides] if args.overrides else {}
+        display_config = (
+            self.shell.user_ns[args.display] if args.display not in [True, None] else {}
+        )
+
+        # Decision: write to file before trying to build and execute Driver
+        # See argument `help` for behavior details
+        if args.write_to_file:
+            if isinstance(args.write_to_file, str):
+                file_path = Path(args.write_to_file)
+            else:
+                file_path = Path(f"{module_name}.py")
+            file_path.write_text(cell)
```

Comment:
this should probably happen after creating the module to catch syntax/import errors, right? Else if it's here bad code could be saved when we should error instead?

Reply:
## 1. Invalid Python
Invalid Python syntax won't get there because of the first line of the magic is `self.shell.ex(cell)`. This will catch any Python error

## Invalid Hamilton
The design decision here is to write the code cell after magic argument parsing, but before further magic logic.

The "contract" of `--write_to_file` is that the code cell is in sync with the module. The user is responsible for what they write to a file (they could write incorrect Python to a `.py` file too).